mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,964 @@
+  SmallVector<Operation *, 8> loadAndStoreOps;
+  long double minCost = LONG_MAX;
+
+  // Get all affine.load and affine.store ops.
+  getAllLoadStores(loopNest[0], loadAndStoreOps);
+
+  // Get size of elements in each memref access. Later to be used
+  // in building reference groups.
+  getElementsSize(loopNest[0]);
+
+  // Get affine access matrices for all load/stores.
+  getAffineAccessMatrices(loopNest[0], loadAndStoreOps, indVarIdMap,
+                          loopNest.size());
+
+  // Get loop carried dependency vector for all affine.for ops in the loop nest.
+  // This is useful in calculating the parallelism cost for each permutation.
+  getAffineLoopCarriedDependencies(loopNest[0], loadAndStoreOps,
+                                   loopNest.size());
+
+  // Get locality information for each affine.forop in the loop nest. This will
+  // be useful in calculating spatial locality cost for each permutation.
+  getCacheLineAccessCounts(loopNest, loadAndStoreOps);
+
+  // Get all permutations which do not violate any dependency constraints.
+  SmallVector<SmallVector<unsigned, 4>, 8> validPerms;
+  getAllValidPermutations(loopNest[0], loopNest.size(),
+                          validPerms);
+
+  // Return size 0 SmallVector to indicate that no interchange exists.
+  if (validPerms.size() < 2)
+    return bestPerm;
+
+  for (auto &perm : validPerms) {
+    int64_t parallelCost =
+        getParallelismCost(perm, this->loopCarriedDependenceVector,
+                           this->loopIterationCounts);
+    long double spatialCost = getSpatialLocalityCost(perm, loopNest,
+        this->numberOfCacheLinesAccessed,
+        this->loopIterationCounts);
+    long double temporalCost = getTemporalReuseCost(perm);
+
+    // Assumption: costs due to parallelism (synchronization) are 100x
+    // more expensive than those due to locality.
+    long double cost =
+        100 * parallelCost + spatialCost + temporalCost;
+
+    if (cost < minCost) {
+      minCost = cost;
+      bestPerm = perm;
+    }
+  }
+
+  //`permuteLoops()` (called in next step) maps loop 'i' to location
+  // bestPerm[i]. But here bestPerm[i] maps to loop
+  // at depth 'i'. Hence, we need to prepare a custom `returnPerm`
+  // according to requirements of `permuteLoops()`.
+  SmallVector<unsigned, 4> returnPerm;
+  returnPerm.resize(bestPerm.size());
+  for (unsigned i = 0; i < bestPerm.size(); i++)
+    returnPerm[bestPerm[i]] = i;
+  return returnPerm;
+}
+
+/// Resets all the protected data members alongwith provided arguments.
+void LoopInterchange::clear(SmallVector<AffineForOp, 4> &loopNest,
+                            DenseMap<Value, unsigned> &indVarIdMap) {
+  loopNest.clear();
+  indVarIdMap.clear();
+  this->loopIterationCounts.clear();
+  this->loopCarriedDependenceVector.clear();
+  this->numberOfCacheLinesAccessed.clear();
+  this->referenceGroups.clear();
+  this->elementSize.clear();
+  this->loopAccessMatrices.clear();
+  this->constVector.clear();
```

Comment:
`this->` is never used to access member variables in the LLVM style.

Reply:
Removed